File: port/java/src/com/gams/utility/Region.java
/*********************************************************************
 * Copyright (c) 2013-2015 Carnegie Mellon University. All Rights Reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following acknowledgments and disclaimers.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 * 
 * 3. The names "Carnegie Mellon University," "SEI" and/or
 * "Software Engineering Institute" shall not be used to endorse or promote
 * products derived from this software without prior written permission. For
 * written permission, please contact permission@sei.cmu.edu.
 * 
 * 4. Products derived from this software may not be called "SEI" nor may "SEI"
 * appear in their names without prior written permission of
 * permission@sei.cmu.edu.
 *
 * 5. Redistributions of any form whatsoever must retain the following
 * acknowledgment:
 *
 * This material is based upon work funded and supported by the Department of
 * Defense under Contract No. FA8721-05-C-0003 with Carnegie Mellon University
 * for the operation of the Software Engineering Institute, a federally funded
 * research and development center. Any opinions, findings and conclusions or
 * recommendations expressed in this material are those of the author(s) and
 * do not necessarily reflect the views of the United States Department of
 * Defense.
 * 
 * NO WARRANTY. THIS CARNEGIE MELLON UNIVERSITY AND SOFTWARE ENGINEERING
 * INSTITUTE MATERIAL IS FURNISHED ON AN "AS-IS" BASIS. CARNEGIE MELLON
 * UNIVERSITY MAKES NO WARRANTIES OF ANY KIND, EITHER EXPRESSED OR IMPLIED,
 * AS TO ANY MATTER INCLUDING, BUT NOT LIMITED TO, WARRANTY OF FITNESS FOR
 * PURPOSE OR MERCHANTABILITY, EXCLUSIVITY, OR RESULTS OBTAINED FROM USE OF THE
 * MATERIAL. CARNEGIE MELLON UNIVERSITY DOES NOT MAKE ANY WARRANTY OF ANY KIND
 * WITH RESPECT TO FREEDOM FROM PATENT, TRADEMARK, OR COPYRIGHT INFRINGEMENT.
 * 
 * This material has been approved for public release and unlimited
 * distribution.
 * 
 * @author James Edmondson <jedmondson@gmail.com>
 *********************************************************************/
package com.gams.utility;

import com.gams.GamsJNI;
import com.madara.containers.StringVector;
import com.madara.KnowledgeBase;

/**
 * A utility class that acts as a facade for a region
 **/
public class Region extends GamsJNI
{
  private native long jni_Region();
  private static native void jni_freeRegion(long cptr);
  private native String jni_getName(long cptr);
  private native void jni_setName(long cptr, String name);
  private native void jni_fromContainer(long cptr, long kb, String name);
  private native void jni_toContainer(long cptr, long kb, String name);
  private native void jni_modify(long cptr);
  private native java.lang.String jni_toString(long cptr);
  private native void jni_addGpsVertex(long cptr, long coord); 
  private native long[] jni_getVertices(long cptr); 
  private native double jni_getArea(long cptr); 
  private native long jni_getBoundingBox(long cptr); 
  private native boolean jni_containsGps(long cptr, long coord);  
  private native double jni_getGpsDistance(long cptr, long coord); 
  private native double jni_getMaxAlt(long cptr); 
  private native double jni_getMinAlt(long cptr); 
  private native double jni_getMaxLat(long cptr); 
  private native double jni_getMinLat(long cptr); 
  private native double jni_getMaxLong(long cptr); 
  private native double jni_getMinLong(long cptr); 
  
  private boolean manageMemory = true;

  /**
   * Default constructor
   **/
  public Region()
  {
    setCPtr(jni_Region());
  }

  /**
   * Converts the position into a string
   * @return position as a string
   **/
  public java.lang.String toString()
  {
    return jni_toString(getCPtr());
  }

  /**
   * Get name of the region
   * @return name of the region
   **/
  public String getName()
  {
    return jni_getName(getCPtr());
  }

  /**
   * Set name of the region
   * @param n     new name of the region
   **/
  public void setName(String n)
  {
    jni_setName(getCPtr(), n);
  }

  /**
   * Adds a GPS vertex to this region
   * @param   point     point to check
   **/
  public void addVertex(GpsPosition point)
  {
    jni_addGpsVertex(getCPtr(),point.getCPtr());
  }
  
  /**
   * Gets area of the region
   * @return area of this region
   **/
  public double getArea()
  {
    return jni_getArea(getCPtr()); 
  }
  
  /**
   * Gets bounding box
   * @return Region object corresponding to bounding box
   **/
  public Region getBoundingBox()
  {
    Region result = new Region();
    result.setCPtr(jni_getBoundingBox(getCPtr()));
    return result;
  }
  
  /**
   * Checks to see if the point is contained in this region
   * @param   point     point to check
   * @return 0 if in region, otherwise distance from region
   **/
  public boolean contains(GpsPosition point)
  {
    return jni_containsGps(getCPtr(),point.getCPtr());
  } 
  
  /**
   * Get distance from any point in this region
   * @param   point     point to check
   * @return 0 if in region, otherwise distance from region
   **/
  public double getDistance(GpsPosition point)
  {
    return jni_getGpsDistance(getCPtr(),point.getCPtr());
  }
  
  /**
   * Gets the GPS vertices that form the region boundary
   * @return vertices that form the region boundary
   **/
  public GpsPosition[] getVertices()
  {
    long[] vertices = jni_getVertices(getCPtr());
    GpsPosition[] result = new GpsPosition[vertices.length];
    
    if(vertices.length > 0)
    {
      result = new GpsPosition[vertices.length];
      for (int i = 0; i < vertices.length; ++i)
        result[i].fromPointer(vertices[i]);
    }
    return result;
  }
  
  /**
   * Gets the maximum altitude
   * @return maximum altitude in the region
   **/
  public double getMaxAlt()
  {
    return jni_getMaxAlt(getCPtr());
  }
  
  /**
   * Gets the minimum altitude
   * @return minimum altitude in the region
   **/
  public double getMinAlt()
  {
    return jni_getMinAlt(getCPtr());
  }
    
  /**
   * Gets the maximum latitude
   * @return maximum latitude in the region
   **/
  public double getMaxLat()
  {
    return jni_getMaxLat(getCPtr());
  }
      
  /**
   * Gets the minimum latitude
   * @return minimum latitude in the region
   **/
  public double getMinLat()
  {
    return jni_getMinLat(getCPtr());
  }
      
  /**
   * Gets the maximum longitude
   * @return maximum longitude in the region
   **/
  public double getMaxLong()
  {
    return jni_getMaxLong(getCPtr());
  }
      
  /**
   * Gets the minimum longitude
   * @return minimum longitude in the region
   **/
  public double getMinLong()
  {
    return jni_getMinLong(getCPtr());
  }
  
  /**
   * Creates a java object instance from a C/C++ pointer
   *
   * @param cptr C pointer to the object
   * @return a new java instance of the underlying pointer
   */
  public static Region fromPointer(long cptr)
  {
    Region ret = new Region();
    ret.manageMemory = true;
    ret.setCPtr(cptr);
    return ret;
  }
  
  /**
   * Creates a java object instance from a C/C++ pointer
   *
   * @param cptr C pointer to the object
   * @param shouldManage  if true, manage the pointer
   * @return a new java instance of the underlying pointer
   */
  public static Region fromPointer(long cptr, boolean shouldManage)
  {
    Region ret = new Region();
    ret.manageMemory=shouldManage;
    ret.setCPtr(cptr);
    return ret;
  }

  /**
   * Helper function for copying values from a MADARA knowledge base
   * @param kb      KnowledgeBase to copy region information to
   **/
  public void fromContainer(com.madara.KnowledgeBase kb)
  {
    fromContainer (kb, getName());
  }

  /**
   * Helper function for copying values from a MADARA knowledge base
   * @param kb      KnowledgeBase to copy region information to
   * @param name    name of the region
   **/
  public void fromContainer(com.madara.KnowledgeBase kb, String name)
  {
    jni_fromContainer(getCPtr(), kb.getCPtr(), name);
  }

  /**
   * Helper function for copying values to a MADARA knowledge base
   * @param kb      KnowledgeBase with region information
   **/
  public void toContainer(com.madara.KnowledgeBase kb)
  {
    toContainer(kb, getName());
  }

  /**
   * Helper function for copying values to a MADARA knowledge base
   * @param kb      KnowledgeBase with region information
   * @param name    name of the region
   **/
  public void toContainer(com.madara.KnowledgeBase kb, String name)
  {
    jni_toContainer(getCPtr(), kb.getCPtr(), name);
  }

  /**
   * Calls toContainer with previous values
   **/
  public void modify()
  {
    jni_modify(getCPtr());
  }

  /**
   * Deletes the C instantiation. To prevent memory leaks, this <b>must</b> be
   * called before an instance of WaitSettings gets garbage collected
   */
  public void free()
  {
    if(manageMemory)
    {
      jni_freeRegion(getCPtr());
    }
  }
  
  /**
   * Cleans up underlying C resources
   * @throws Throwable necessary for override but unused
   */
  @Override
  protected void finalize() throws Throwable
  {
    try {
      free();
    } catch (Throwable t) {
      throw t;
    } finally {
      super.finalize();
    }
  }
}

File: src/gams/utility/Region.cpp
/**
 * Copyright (c) 2014 Carnegie Mellon University. All Rights Reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * 
 * 1. Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following acknowledgments and disclaimers.
 * 
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 * 
 * 3. The names "Carnegie Mellon University," "SEI" and/or "Software
 *    Engineering Institute" shall not be used to endorse or promote products
 *    derived from this software without prior written permission. For written
 *    permission, please contact permission@sei.cmu.edu.
 * 
 * 4. Products derived from this software may not be called "SEI" nor may "SEI"
 *    appear in their names without prior written permission of
 *    permission@sei.cmu.edu.
 * 
 * 5. Redistributions of any form whatsoever must retain the following
 *    acknowledgment:
 * 
 *      This material is based upon work funded and supported by the Department
 *      of Defense under Contract No. FA8721-05-C-0003 with Carnegie Mellon
 *      University for the operation of the Software Engineering Institute, a
 *      federally funded research and development center. Any opinions,
 *      findings and conclusions or recommendations expressed in this material
 *      are those of the author(s) and do not necessarily reflect the views of
 *      the United States Department of Defense.
 * 
 *      NO WARRANTY. THIS CARNEGIE MELLON UNIVERSITY AND SOFTWARE ENGINEERING
 *      INSTITUTE MATERIAL IS FURNISHED ON AN "AS-IS" BASIS. CARNEGIE MELLON
 *      UNIVERSITY MAKES NO WARRANTIES OF ANY KIND, EITHER EXPRESSED OR
 *      IMPLIED, AS TO ANY MATTER INCLUDING, BUT NOT LIMITED TO, WARRANTY OF
 *      FITNESS FOR PURPOSE OR MERCHANTABILITY, EXCLUSIVITY, OR RESULTS
 *      OBTAINED FROM USE OF THE MATERIAL. CARNEGIE MELLON UNIVERSITY DOES
 *      NOT MAKE ANY WARRANTY OF ANY KIND WITH RESPECT TO FREEDOM FROM PATENT,
 *      TRADEMARK, OR COPYRIGHT INFRINGEMENT.
 * 
 *      This material has been approved for public release and unlimited
 *      distribution.
 **/

/**
 * @file Region.cpp
 * @author James Edmondson <jedmondson@gmail.com>
 *
 * This file contains a utility class for working with regions
 **/

#include <sstream>
#include <string>
#include <vector>
#include <cmath>

#include "Region.h"
#include "madara/utility/Utility.h"
#include "gams/loggers/Global_Logger.h"

#include "madara/knowledge_engine/containers/Integer.h"
#include "madara/knowledge_engine/containers/Vector.h"

using std::string;
using std::stringstream;
using std::vector;

typedef  Madara::Knowledge_Record::Integer Integer;

gams::utility::Region::Region (
  const std::vector <GPS_Position> & init_vertices, unsigned int t, 
  const std::string& name) :
  vertices (init_vertices), name_(name), type_ (t)
{
  calculate_bounding_box ();
}

gams::utility::Region::~Region ()
{
}

void
gams::utility::Region::operator= (const Region& rhs)
{
  if (this != &rhs)
  {
    this->vertices = rhs.vertices;
    this->name_ = rhs.name_;
    this->type_ = rhs.type_;
    calculate_bounding_box ();
  }
}

bool
gams::utility::Region::operator== (const Region& rhs) const
{
  if (this == &rhs)
    return true;

  if (vertices.size () != rhs.vertices.size ())
    return false;

  // ensure all contents are the same
  for (size_t i = 0; i < vertices.size (); ++i)
  {
    size_t j;
    for (j = 0; j < vertices.size (); ++j)
      if (vertices[i] == rhs.vertices[j])
        break;
    if (j == vertices.size())
      return false;
  }

  return true;
}

bool
gams::utility::Region::operator!= (const Region& rhs) const
{
  return !(*this == rhs);
}

std::string
gams::utility::Region::get_name () const
{
  return name_;
}

void
gams::utility::Region::set_name (const std::string& n)
{
  name_ = n;
}

bool
gams::utility::Region::contains (const GPS_Position & p) const
{
  // check if in bounding box
  if (p.latitude () < min_lat_ || p.latitude () > max_lat_ ||
      p.longitude () < min_lon_ || p.longitude () > max_lon_)
  {
    return false;
  }

  // check if point in polygon code from 
  // http://www.ecse.rpi.edu/Homepages/wrf/Research/Short_Notes/pnpoly.html
  size_t i, j;
  bool ret = false;
  for (i = 0, j = vertices.size() - 1; i < vertices.size(); j = i++)
  {
    if ( (vertices[i].longitude () > p.longitude ()) !=
        (vertices[j].longitude () > p.longitude ()))
    {
      if (p.latitude () < (vertices[j].latitude () - vertices[i].latitude ()) * (p.longitude () - vertices[i].longitude ()) / 
                (vertices[j].longitude () - vertices[i].longitude ()) + 
                 vertices[i].latitude ())
      {
        ret = !ret;
      }
    }
  }

  // check if this is a vertex point
  if (!ret)
  {
    for (unsigned int i = 0; i < vertices.size() && !ret; ++i)
      ret = (vertices[i] == p);
  }

  // TODO: add check for border point

  return ret;
}

double
gams::utility::Region::distance (const GPS_Position& p) const
{
  // if point is in region, then the distance is 0
  if (contains (p))
    return 0;

  // convert to cartesian coords with equirectangular projection
  const GPS_Position sw (min_lat_, min_lon_);
  vector<Position> local_vertices;
  for (size_t i = 0; i < vertices.size (); ++i)
    local_vertices.push_back (vertices[i].to_position (sw));
  Position local_p = p.to_position (sw);

  // else we check for distance from each edge
  double min_dist = DBL_MAX;
  for (size_t i = 0; i < local_vertices.size (); ++i)
  {
    const size_t i_1 = (i + 1) % local_vertices.size();
    double dist = local_vertices[i].distance_to_2d (local_vertices[i_1], local_p);
    if (dist < min_dist)
      min_dist = dist;
  }

  return min_dist;
}

bool
gams::utility::Region::contains (const Position & p,
  const GPS_Position & ref) const
{
  // convert to GPS_Position and then just use other contains
  return contains (utility::GPS_Position::to_gps_position (p, ref));
}

gams::utility::Region
gams::utility::Region::get_bounding_box () const
{
  Region ret;

  GPS_Position p;

  p.latitude (min_lat_);
  p.longitude (min_lon_);
  p.altitude (0);
  ret.vertices.push_back (p);

  p.latitude (min_lat_);
  p.longitude (max_lon_);
  p.altitude (0);
  ret.vertices.push_back (p);

  p.latitude (max_lat_);
  p.longitude (max_lon_);
  p.altitude (0);
  ret.vertices.push_back (p);

  p.latitude (max_lat_);
  p.longitude (min_lon_);
  p.altitude (0);
  ret.vertices.push_back (p);

  ret.min_lat_ = this->min_lat_;
  ret.max_lat_ = this->max_lat_;
  ret.min_lon_ = this->min_lon_;
  ret.max_lon_ = this->max_lon_;
  ret.min_alt_ = this->min_alt_;
  ret.max_alt_ = this->max_alt_;

  return ret;
}

double
gams::utility::Region::get_area () const
{
  if (vertices.size() < 3)
    return 0; // degenerate polygon

  // see http://geomalgorithms.com/a01-_area.html
  // Convert all units to cartesian
  vector<Position> cart_vertices;
  for (unsigned int i = 0; i < vertices.size(); ++i)
  {
    Position p = vertices[i].to_position (vertices[0]);
    cart_vertices.push_back (p);
  }

  // perform calculations with cartesian vertices
  double area = 0.0;
  size_t i, j, k;
  size_t num_vertices = cart_vertices.size ();
  for (i = 1, j = 2, k = 0; i < num_vertices; ++i, ++j, ++k)
  {
    area += cart_vertices[i].x *
      (cart_vertices[j % num_vertices].y - cart_vertices[k].y);
  }
  area += cart_vertices[0].x * (cart_vertices[1].y - cart_vertices[num_vertices - 1].y);
  return fabs(area / 2);
}

string
gams::utility::Region::to_string (const string & delimiter) const
{
  stringstream buffer;

  if (vertices.size () > 0)
  {
    buffer << vertices[0].to_string ();
    for (unsigned int i = 1; i < vertices.size (); ++i)
    {
      buffer << delimiter;
      buffer << vertices[i].to_string ();
    }
  }

  return buffer.str ();
}

void
gams::utility::Region::to_container (
  Madara::Knowledge_Engine::Knowledge_Base& kb)
{
  to_container (kb, get_name ());
}

void
gams::utility::Region::to_container (
  Madara::Knowledge_Engine::Knowledge_Base& kb, const std::string& name) 
{
  // set object type
  Madara::Knowledge_Engine::Containers::Integer obj_type (
    name + object_type_suffix_, kb);
  obj_type = REGION_TYPE_ID;
  
  // set type of region
  Madara::Knowledge_Engine::Containers::Integer type (name + ".type", kb);
  type = type_;

  switch (type.to_integer ())
  {
    case 0: // arbitrary convex polygon
    {
      // set size
      Madara::Knowledge_Engine::Containers::Integer size (name + ".size", kb);
      size = vertices.size();
    
      // set vertices
      Madara::Knowledge_Engine::Containers::Native_Double_Vector target;
      for (unsigned int i = 0; i < vertices.size (); ++i)
      {
        std::stringstream vert_name;
        vert_name << name << "." << i;
        target.set_name (vert_name.str (), kb);
        target.set (2, vertices[i].z);
        target.set (0, vertices[i].x);
        target.set (1, vertices[i].y);
      }

      break;
    }
    default:
      madara_logger_ptr_log (gams::loggers::global_logger.get (),
        gams::loggers::LOG_ERROR,
         "gams::utility::Region::to_container:" \
        " ERROR: invalid region type %" PRId64 ".\n", type.to_integer ());
  }
}

bool
gams::utility::Region::from_container (
  Madara::Knowledge_Engine::Knowledge_Base& kb)
{
  return from_container (kb, get_name ());
}

bool
gams::utility::Region::from_container (
  Madara::Knowledge_Engine::Knowledge_Base& kb, const std::string& name)
{
  if (!check_valid_type (kb, name))
  {
    madara_logger_ptr_log (gams::loggers::global_logger.get (),
      gams::loggers::LOG_ERROR,
      "gams::utility::Region::from_container:" \
      " \"%s\" is not a valid Region\n", name.c_str ());
    return false;
  }

  madara_logger_ptr_log (gams::loggers::global_logger.get (),
    gams::loggers::LOG_DETAILED,
    "gams::utility::Region::from_container:" \
    " name = %s\n", name.c_str ());

  // get type
  Madara::Knowledge_Engine::Containers::Integer type (name + ".type", kb);
  type_ = type.to_integer ();

  // set name if necessary
  if (name_ == "")
    name_ = name;

  // get vertices
  switch (type_)
  {
    case 0: // arbitrary convex polygon
    {
      madara_logger_ptr_log (gams::loggers::global_logger.get (),
        gams::loggers::LOG_DETAILED,
         "gams::utility::Region::from_container:" \
         " type is arbitrary convex polygon\n");

      // get size
      Madara::Knowledge_Engine::Containers::Vector vertices_knowledge;
      vertices_knowledge.set_name (name, kb);
      vertices_knowledge.resize();
      unsigned int size = vertices_knowledge.size ();
      vertices.resize (size);

      madara_logger_ptr_log (gams::loggers::global_logger.get (),
        gams::loggers::LOG_DETAILED,
        "gams::utility::Region::from_container:" \
        " size is %u\n", size);

      // get each of the vertices
      for (Integer i = 0; i < size; ++i)
      {
        std::vector<double> coords (vertices_knowledge[i].to_doubles ());

        if (coords.size () == 2)
        {
          madara_logger_ptr_log (gams::loggers::global_logger.get (),
            gams::loggers::LOG_DETAILED,
            "gams::utility::Region::from_container:" \
            " Adding coordinate (%f, %f)\n",
            coords[0], coords[1]);
          vertices[i] = GPS_Position(coords[0], coords[1]);
        }
        else if (coords.size () == 3)
        {
          madara_logger_ptr_log (gams::loggers::global_logger.get (),
            gams::loggers::LOG_DETAILED,
            "gams::utility::Region::from_container:" \
            " Adding coordinate (%f, %f, %f)\n",
            coords[0], coords[1], coords[2]);
          vertices[i] = GPS_Position(coords[0], coords[1], coords[2]);
        }
        else
        {
          madara_logger_ptr_log (gams::loggers::global_logger.get (),
            gams::loggers::LOG_ERROR,
            "gams::utility::Region::from_container:" \
            " ERROR: invalid coordinate type at %s.%u\n", name.c_str (), i);
        }
      }

      calculate_bounding_box ();

      break;
    }
    default:
    {
      madara_logger_ptr_log (gams::loggers::global_logger.get (),
        gams::loggers::LOG_ERROR,
        "gams::utility::Region::from_container:" \
        " ERROR: invalid region type %" PRId64 ".\n", type_);
      return false;
    }
  }

  return true;
}

void
gams::utility::Region::calculate_bounding_box ()
{
  min_lat_ = DBL_MAX;
  min_lon_ = DBL_MAX;
  min_alt_ = DBL_MAX;
  max_lat_ = -DBL_MAX;
  max_lon_ = -DBL_MAX;
  max_alt_ = -DBL_MAX;
  for (unsigned int i = 0; i < vertices.size(); ++i)
  {
    min_lat_ = (min_lat_ > vertices[i].latitude ()) ?
      vertices[i].latitude () : min_lat_;
    min_lon_ = (min_lon_ > vertices[i].longitude ()) ?
      vertices[i].longitude () : min_lon_;
    min_alt_ = (min_alt_ > vertices[i].altitude ()) ?
      vertices[i].altitude () : min_alt_;

    max_lat_ = (max_lat_ < vertices[i].latitude ()) ?
      vertices[i].latitude () : max_lat_;
    max_lon_ = (max_lon_ < vertices[i].longitude ()) ?
      vertices[i].longitude () : max_lon_;
    max_alt_ = (max_alt_ < vertices[i].altitude ()) ?
      vertices[i].altitude () : max_alt_;
  }
}

bool
gams::utility::Region::check_valid_type (
  Madara::Knowledge_Engine::Knowledge_Base& kb, const std::string& name) const
{
  return Containerize::is_valid_type (kb, name, REGION_TYPE_ID);
}
